python/plugins/processing/algs/qgis/Dissolve.py

```diff
@@ -49,17 +50,24 @@ class Dissolve(GeoAlgorithm):
 
     def processAlgorithm(self, progress):
         useField = not self.getParameterValue(Dissolve.DISSOLVE_ALL)
-        fieldname = self.getParameterValue(Dissolve.FIELD)
         vlayerA = dataobjects.getObjectFromUri(
             self.getParameterValue(Dissolve.INPUT))
-        field = vlayerA.fieldNameIndex(fieldname)
         vproviderA = vlayerA.dataProvider()
-        fields = vproviderA.fields()
+        fields = QgsFields()
+        if useField:
+            fieldname = self.getParameterValue(Dissolve.FIELD)
+            fieldIdx = vlayerA.fieldNameIndex(fieldname)
+            field = vlayerA.fields()[fieldIdx]
```

Comment:
Stability is increased by not mixing *provider* and *layer* indexes which may differ.

And I would say it increases readability.